daemon/containerd/image_list.go

```diff
@@ -98,16 +98,42 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 	contentStore := i.client.ContentStore()
 	uniqueImages := map[digest.Digest]images.Image{}
 	tagsByDigest := map[digest.Digest][]string{}
+	intermediateImages := map[digest.Digest]struct{}{}
+
+	hideIntermediate := !opts.All
+	if hideIntermediate {
+		for _, img := range imgs {
+			parent, ok := img.Labels[imageLabelClassicBuilderParent]
```

Comment:
We should be doing this only for dangling images. Also, why do we create a map here and not do this in the main loop?

Reply:
> We should be doing this only for dangling images

Nope! The label points to the parent. If we'd do this only for dangling images we'd miss the layer just before the final one.

Consider the following "family tree":
```
busybox                        # FROM busybox
|
dangling-X (parent: busybox).  # ADD foo /bar
|
asdf (parent: dangling-B)      # CMD ["cat", "/bar"]
```

Reply:
> why do we create a map here and not do this in the main loop

This allows us to skip the intermediate images immediately based on the data we already have (labels in the containerd image object) before calling `filter` which is much heavier call.

We can't do this without the "extra pass" because parent label on the X doesn't tell us if X is intermediate or not. We can only determine if image X is intermediate if we know that there's other image that claims to be its child.

Reply:
Gotcha, I didn't completely understand the code before making that comment, sorry :)

Reply:
No problem, if it wasn't obvious by looking at it, it's definitely worth for it to explain it :)